### ISSUE TITLE
Sugestion: add private repositories support in recipe engine

### ISSUE THREAD

**Issue by SGTGunner** [citizenfx/txAdmin#349]
The following use case: As a server developer, I would like to supply the Github OAuth token to allow for private GitHub repos to be downloaded as part of the recipe.

There is a way to do it today directly via the command line. To also have it be exposed in a recipe.

`git pull https://$OAUTH_TOKEN:x-oauth-basic@github.com/name/repo.git master` 

It could be entered in the UI and stored as a tmp variable that then can be defined in the yaml?

**Comment by tabarra** [comment]
Right now we are not using git, just downloading the file with the following URL `https://api.github.com/repos/${repoOwner}/${repoName}/zipball/${reference}`.  
If you find a way to do that without git, please do a PR to the `recipeEngine.js` file.

**Comment by SGTGunner** [comment]
Oh, okay. Thanks for the reply! I will take a look at creating a PR.

**Comment by tabarra** [comment]
Maybe github allows oauth token in the HTTP request as the Authorization header?! Who knows

**Comment by SGTGunner** [comment]
yeah, that is what I am going to look at. The big question is what would you recommend as a way to pass the token? By the UI or through a var? Obviously, I don't want to store it in the recipe :-)

**Comment by tabarra** [comment]
for it to show on the UI, you have to create a var for it, so the var could have the token, or the var could be empty and the user fills it manually

**Comment by varnem2** [comment]
Digging up this issue, I'm going to share some thoughts here.


## Background

As a review for the download_github task in the recipeEngine.js it seems as though it's making 2 calls to github.com. (It seems this was refreshed from https://github.com/axios/axios over to https://github.com/sindresorhus/got between the time I've looked at this las.)

The first call occurs at https://github.com/tabarra/txAdmin/blob/63afc1048806d251669419fb6311ef506dbf672a/src/extras/recipeEngine.js#L113 where we get the default branch name for the repo to later use here https://github.com/tabarra/txAdmin/blob/63afc1048806d251669419fb6311ef506dbf672a/src/extras/recipeEngine.js#L122 .

Both of these calls would need authenticated through Github for the code to continue to work as it does today and expand it to a private repo. 

Creating a personal access token in github as the instructions indicated [here](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) (We'll define this token as GITHUB_PAT going forward.) allows a user to make both of these calls against an "internal" repo or a "private" repo. Upon further experimentation you'll need to add a header to both of the indicated calls above following this object. { "key": "Authorization", "value": `Bearer ${GITHUB_PAT}` }.

Passing in the header into the call where GITHUB_PAT equals an empty or a null value causes 401s to be returned from github.com regardless of the permission/visibility settings of the repo. The existence of the header does not limit the ability to download 'public' repos. The addition of this GITHUB_PAT token may cause tighter rate limiting than without any token at all, however that does still need to be experimented.

## What I need help with is design and a few questions.

- What do you want the scope of this GITHUB_PAT to be? Recipe 1 or Recipe 2? Recipe 1 is an example of a global scope. While Recipe 2 would be an example of the scope being limited to the task?

Recipe 1

```(yaml)
$engine: 3
$onesync: on
name: QBCore Framework RedM
version: 0.1.0
author: The QBCore Family
description: An advanced RedM framework including jobs, gangs, housing & more!
github_personal_access_token: ADSFADSADF

tasks:
  # Download Base Files
  - action: download_github
    src: https://github.com/qbcore-redm-framework/txAdminRecipe
    ref: main
    dest: ./tmp/qbrcore
```

Recipe 2
```(yaml)
$engine: 3
$onesync: on
name: QBCore Framework RedM
version: 0.1.0
author: The QBCore Family
description: An advanced RedM framework including jobs, gangs, housing & more!

tasks:
  # Download Base Files
  - action: download_github
    src: https://github.com/qbcore-redm-framework/txAdminRecipe
    ref: main
    dest: ./tmp/qbrcore
    github_personal_access_token: ADSFADSADF
```

- Would we want to check the visibility of the repo to decide if we add the GITHUB_PAT token to the call? That would add a 3rd additional call to github.com for each github_download task.

- Would you want to do an [implicit Auth](https://github.com/auth0/docs/blob/master/articles/api-auth/tutorials/implicit-grant.md) where txAdmin would create a GITHUB_PAT for the user as the user uses the UI?

- Would you prefer a dark launch of the GITHUB_PAT in multiple PRs or just build out the whole thing?

**Comment by tabarra** [comment]
@varnem2 Thank you for the analysis, and sorry for taking 2 years to reply :P  
As per the [docs](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28), to looks like personal auth token will grant 5k reqs/hour.  

## My proposal:
1. Recipes with private repos, can flag the recipe with the meta tag `$requiresGithubToken`.
2. For the token above, txAdmin would first try to retrieve it from the `TXADMIN_GITHUB_TOKEN` env var, and also allow it to be typed in the deployer page, required if the env var is not found.
3. That toke would be used for all requests from the `download_github` action.
4. It would not be possible to configure a token for a specific request, they all gonna use the same token.
5. The shared hosting providers (usually pterodactyl) could add their own token to solve the deployer rate limiting issues that are too frequent.  

What do you guys think?

**Comment by bsk193** [comment]
> @varnem2 Thank you for the analysis, and sorry for taking 2 years to reply :P As per the [docs](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28), to looks like personal auth token will grant 5k reqs/hour.
> 
> ## My proposal:
> 1. Recipes with private repos, can flag the recipe with the meta tag `$requiresGithubToken`.
> 2. For the token above, txAdmin would first try to retrieve it from the `TXADMIN_GITHUB_TOKEN` env var, and also allow it to be typed in the deployer page, required if the env var is not found.
> 3. That toke would be used for all requests from the `download_github` action.
> 4. It would not be possible to configure a token for a specific request, they all gonna use the same token.
> 5. The shared hosting providers (usually pterodactyl) could add their own token to solve the deployer rate limiting issues that are too frequent.
> 
> What do you guys think?


Lets do it!
